File: src/core/systems/timetable-generator/data-structures/PersistentHeap.cpp
#include <emscripten/bind.h>

#include <algorithm>
#include <functional>
#include <random>
#include <vector>

class PhNode {
 private:
  int curr;             // 현재 정점
  int next;             // 다음 정점
  double sidetrack;     // 대체 경로 비용
  PhNode *children[2];  // 자식 힙

 public:
  PhNode(int _curr, int _next, double _sidetrack, PhNode *_left, PhNode *_right)
      : curr(_curr), next(_next), sidetrack(_sidetrack) {
    children[0] = _left;
    children[1] = _right;
  }

  int getCurrVertex() { return curr; }
  int getNextVertex() { return next; }
  double getSideTrack() { return sidetrack; }
  PhNode getChild(int _nth) {
    return children[_nth] == nullptr ? PhNode(-1, -1, -1, nullptr, nullptr)
                                     : *children[_nth];
  }

  // 파라미터로 받은 node를 복사
  static PhNode *copy(PhNode *node) {
    if (node == (PhNode *)nullptr) return (PhNode *)nullptr;
    return new PhNode(node->curr, node->next, node->sidetrack,
                      node->children[0], node->children[1]);
  }

  // 두 노드 합치기
  static PhNode *merge(PhNode *fst, PhNode *snd,
                       std::function<int(int, int)> &randomInt) {
    if (fst == (PhNode *)nullptr) return snd;
    if (snd == (PhNode *)nullptr) return fst;
    if (fst->sidetrack > snd->sidetrack) std::swap(fst, snd);
    int idx = randomInt(0, 1);
    if (fst->children[idx] != (PhNode *)nullptr)
      fst->children[idx] = PhNode::copy(fst->children[idx]);
    fst->children[idx] = PhNode::merge(fst->children[idx], snd, randomInt);
    return fst;
  }
};

class PersistentHeap {
 private:
  int tableSize;
  std::vector<PhNode *> heapTable;
  std::mt19937 randomGen;

 public:
  PersistentHeap(int _tableSize) : randomGen(0x69420) { assign(_tableSize); }

  void assign(int _tableSize) {
    tableSize = _tableSize;
    heapTable.assign(_tableSize, nullptr);
  }

  PhNode getHeap(int _idx) {
    return heapTable[_idx] == nullptr ? PhNode(-1, -1, -1, nullptr, nullptr)
                                      : *heapTable[_idx];
  }

  // PhNode 복사
  void copyTo(int _idx, int _desIdx) {
    heapTable[_desIdx] = PhNode::copy(heapTable[_idx]);
  }

  // 새 PhNode 삽입
  void insertAt(int _idx, int _curr, int _next, double _sidetrack) {
    std::function<int(int, int)> randomInt = [&](int lb, int ub) {
      return std::uniform_int_distribution<int>(lb, ub)(randomGen);
    };

    heapTable[_idx] = PhNode::merge(
        heapTable[_idx], new PhNode(_curr, _next, _sidetrack, nullptr, nullptr),
        randomInt);
  }
};

EMSCRIPTEN_BINDINGS(persistentHeap_module) {
  emscripten::class_<PhNode>("PhNode")
      .function("getCurrVertex", &PhNode::getCurrVertex)
      .function("getNextVertex", &PhNode::getNextVertex)
      .function("getSideTrack", &PhNode::getSideTrack)
      .function("getChild", &PhNode::getChild);

  emscripten::class_<PersistentHeap>("PersistentHeap")
      .constructor<int>()
      .function("assign", &PersistentHeap::assign)
      .function("getHeap", &PersistentHeap::getHeap)
      .function("copyTo", &PersistentHeap::copyTo)
      .function("insertAt", &PersistentHeap::insertAt);
}

File: src/core/systems/timetable-generator/data-structures/PersistentHeap.ts
import { RandomGenerator } from "../utility/RandomGenerator"


export class PhNode {
    public son: (PhNode | null)[];
    public value: { curr: number; next: number; sidetrack: number; };

    constructor() {
        this.son = [null, null];
        this.value = { curr: 0, next: 0, sidetrack: 0 };
    }

    public static compare(fst: PhNode, snd: PhNode) {
        if (fst.value.sidetrack === snd.value.sidetrack) return 0;
        return fst.value.sidetrack < snd.value.sidetrack ? -1 : 1;
    }
}


export class PersistentHeap {
    private size: number;
    private heapTable: (PhNode | null)[];

    constructor(size: number = 0) {
        this.size = size;
        this.heapTable = new Array(this.size).fill(null);
    }

    public assign(size: number): void {
        this.size = size;
        this.heapTable = new Array(this.size).fill(null);
    }

    public get(idx: number): (PhNode | null) {
        return this.heapTable[idx];
    }

    public set(idx: number, node: (PhNode | null)): void {
        this.heapTable[idx] = node;
    }

    /** 새로운 Heap 추가 */
    public static newHeap(u: number, v: number, sidetrack: number): PhNode {
        let newNode = new PhNode();
        newNode.value = { curr: u, next: v, sidetrack: sidetrack };
        return newNode;
    }

    /** PhNode 복사 */
    public static copy(node: (PhNode | null)): (PhNode | null) {
        if (node === null) return null;
        const newNode = new PhNode();
        newNode.son[0] = node.son[0];
        newNode.son[1] = node.son[1];
        newNode.value = { curr: node.value.curr, next: node.value.next, sidetrack: node.value.sidetrack };
        return newNode;
    }

    /** 주어진 두 Heap을 합치고, 최상위 노드를(root)를 반환 */
    public static merge(fstNode: (PhNode | null), sndNode: (PhNode | null)): (PhNode | null) {
        if (fstNode === null) return sndNode;
        if (sndNode === null) return fstNode;
        if (PhNode.compare(fstNode, sndNode) > 0) {
            let temp = fstNode;
            fstNode = sndNode;
            sndNode = temp;
        }
        let sonIdx = RandomGenerator.randNumber(0, 1);
        // TODO: 나중에 null오류가 있다면 확인 필요
        if (fstNode.son[sonIdx] !== null) fstNode.son[sonIdx] = PersistentHeap.copy(fstNode.son[sonIdx]);
        fstNode.son[sonIdx] = PersistentHeap.merge(fstNode.son[sonIdx], sndNode);
        return fstNode;
    }
}